src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -0,0 +1,177 @@
+  @Input() container: string;
+  @Input() headerSelectors = '.docs-markdown-h3,.docs-markdown-h4';
+
+  _rootUrl: string;
+  private _scrollContainer: any;
+  private _scrollSubscription: Subscription;
+  private _routeSubscription: Subscription;
+  private _fragmentObserver: MutationObserver;
+  private _headerObserver: MutationObserver;
+
+  private get scrollOffset(): number {
+    const {top} = this._element.nativeElement.getBoundingClientRect();
+    if (typeof this._scrollContainer.scrollTop !== 'undefined') {
+      return this._scrollContainer.scrollTop + top;
+    } else if (typeof this._scrollContainer.pageYOffset !== 'undefined') {
+      return this._scrollContainer.pageYOffset + top;
+    }
+  }
+
+  constructor(private _router: Router,
+              private _route: ActivatedRoute,
+              private _element: ElementRef,
+              @Inject(DOCUMENT) private _document: any) {
+
+    this._routeSubscription = this._router.events.subscribe((event: Event) => {
+      if (event instanceof NavigationEnd) {
+        const rootUrl = _router.url.split('#')[0];
+        if (rootUrl !== this._rootUrl) {
+          this.links.splice(0, this.links.length);
+          this.createLinks();
+          this._rootUrl = rootUrl;
+        }
+      }
+    });
+  }
+
+  ngOnInit(): void {
+    this._scrollContainer = this.container ?
+      this._document.querySelectorAll(this.container)[0] : window;
+
+    this._scrollSubscription = Observable
+      .fromEvent(this._scrollContainer, 'scroll')
+      .debounceTime(10)
+      .subscribe(this.onScroll.bind(this));
+  }
+
+  ngAfterViewInit(): void {
+    this._route.fragment.subscribe(fragment => {
+      // ensure we don't create memory leaks
+      if (this._fragmentObserver) {
+        this._fragmentObserver.disconnect();
+      }
+
+      // create a observer to ensure the element is loaded
+      // into view before trying to query and scroll to it
+      this._fragmentObserver = new MutationObserver(() => {
+        const target = document.getElementById(fragment);
+        if (target) {
+          target.scrollIntoView();
+          this._fragmentObserver.disconnect();
+          this._fragmentObserver = null;
+        }
+      });
+
+      this._fragmentObserver.observe(document.body, {
+        childList: true,
+        subtree: true
+      });
+    });
+  }
+
+  ngOnDestroy(): void {
+    this._routeSubscription.unsubscribe();
+    this._scrollSubscription.unsubscribe();
+
+    if (this._fragmentObserver) {
+      this._fragmentObserver.disconnect();
+    }
+
+    if (this._headerObserver) {
+      this._headerObserver.disconnect();
+    }
+  }
+
+  createLinks(): void {
+    // ensure we don't create memory links
+    if (this._headerObserver) {
+      this._headerObserver.disconnect();
+    }
+
+    // observe body content changes until the headers
+    // are painted. this is required because the page can load async
+    this._headerObserver = new MutationObserver(() => {
```

Comment:
Unfortunately the router events do not give me enough information just to attach one to the route event.

For example, I listen to `this._route.fragment` which returns me the route fragment a user clicked. From that fragment I then ensure its loaded and then scroll into view.

On the `this._router.events`, I'm listening for when the user navigates to a totally different page. For example, navigating from accordion to autocomplete. In this scenario, the page doesn't actually change since its the same route it just updates the content. In this scenario, I want to update the links only.

If the router events gave me more information, I could easily consolidate these but at this time I need 2 different subscriptions and observers to make the appropriate action.